File: python/sol/comm_X.py
# -*- coding: utf-8 -*-
"""
comm_X.py (MPI)
X境界のHy/Hzを共有する
"""

from mpi4py import MPI
from numba import jit, prange

def share(mode, Hy, Hz,
    SendBuf_Bx_hy, SendBuf_Bx_hz, RecvBuf_Bx_hy, RecvBuf_Bx_hz,
    Bx_jhy, Bx_jhz, Bx_khy, Bx_khz,
    Npx, Npy, Npz, Ipx, Ipy, Ipz,
    iMin, iMax, Ni, Nj, Nk, N0):

    assert((mode == 0) or (mode == 1))

    jhy0, jhy1 = Bx_jhy
    khy0, khy1 = Bx_khy
    jhz0, jhz1 = Bx_jhz
    khz0, khz1 = Bx_khz

    bx = [[Ipx > 0, Ipx < Npx - 1], [Ipx == 0, Ipx == Npx - 1]]
    px = [[Ipx - 1, Ipx + 1], [Npx - 1, 0]]
    isend = [iMin + 0, iMax - 1]
    irecv = [iMin - 1, iMax + 0]

    for side in range(2):
        if bx[mode][side]:
            # from H to buffer
            i = isend[side]
            h_to_buffer(i, Hy, SendBuf_Bx_hy, jhy0, jhy1, khy0, khy1, Ni, Nj, Nk, N0)
            h_to_buffer(i, Hz, SendBuf_Bx_hz, jhz0, jhz1, khz0, khz1, Ni, Nj, Nk, N0)

            # MPI
            ipx = px[mode][side]
            dst = (ipx * Npy * Npz) + (Ipy * Npz) + Ipz
            MPI.COMM_WORLD.Sendrecv(SendBuf_Bx_hy, dst, recvbuf=RecvBuf_Bx_hy, source=dst)
            MPI.COMM_WORLD.Sendrecv(SendBuf_Bx_hz, dst, recvbuf=RecvBuf_Bx_hz, source=dst)

            # from buffer to H
            i = irecv[side]
            buffer_to_h(i, Hy, RecvBuf_Bx_hy, jhy0, jhy1, khy0, khy1, Ni, Nj, Nk, N0)
            buffer_to_h(i, Hz, RecvBuf_Bx_hz, jhz0, jhz1, khz0, khz1, Ni, Nj, Nk, N0)

# H to buffer
@jit(cache=True, nopython=True, nogil=True, parallel=True)
def h_to_buffer(i, h, buf, j0, j1, k0, k1, Ni, Nj, Nk, N0):

    for j in prange(j0, j1 + 1):
        for k in range(k0, k1 + 1):
            m = (j - j0) * (k1 - k0 + 1) + (k - k0)
            n = (Ni * i) + (Nj * j) + (Nk * k) + N0
            buf[m] = h[n]

# H to buffer
@jit(cache=True, nopython=True, nogil=True, parallel=True)
def buffer_to_h(i, h, buf, j0, j1, k0, k1, Ni, Nj, Nk, N0):

    for j in prange(j0, j1 + 1):
        for k in range(k0, k1 + 1):
            m = (j - j0) * (k1 - k0 + 1) + (k - k0)
            n = (Ni * i) + (Nj * j) + (Nk * k) + N0
            h[n] = buf[m]

File: mpi/comm_X.c
/*
comm_X.c (MPI)
share X plane Hy and Hz
mode = 0/1 : boundary/PBC
*/

#ifdef _MPI
#include <mpi.h>
#endif

#include "ofd.h"

// H to buffer
static void h_to_buffer(int i, const real_t h[], real_t buf[], const int jrange[], const int krange[])
{
	for (int j = jrange[0]; j <= jrange[1]; j++) {
	for (int k = krange[0]; k <= krange[1]; k++) {
		const int64_t m = (j - jrange[0]) * (krange[1] - krange[0] + 1) + (k - krange[0]);
		buf[m] = h[NA(i, j, k)];
	}
	}
}


// buffer to H
static void buffer_to_h(int i, const real_t buf[], real_t h[], const int jrange[], const int krange[])
{
	for (int j = jrange[0]; j <= jrange[1]; j++) {
	for (int k = krange[0]; k <= krange[1]; k++) {
		const int64_t m = (j - jrange[0]) * (krange[1] - krange[0] + 1) + (k - krange[0]);
		h[NA(i, j, k)] = buf[m];
	}
	}
}


void comm_X(int mode)
{
#ifdef _MPI
	MPI_Status status;
	const int tag = 0;
	int bx[][2] = {{Ipx > 0, Ipx < Npx - 1}, {Ipx == 0, Ipx == Npx - 1}};
	int px[][2] = {{Ipx - 1, Ipx + 1}, {Npx - 1, 0}};
	int irecv[] = {iMin - 1, iMax + 0};
	int isend[] = {iMin + 0, iMax - 1};
	int i;

	// -X/+X boundary
	for (int side = 0; side < 2; side++) {
		if (bx[mode][side]) {
			// H to buffer
			i = isend[side];
			h_to_buffer(i, Hy, Sendbuf_hy_x, Bid.jhy_x, Bid.khy_x);
			h_to_buffer(i, Hz, Sendbuf_hz_x, Bid.jhz_x, Bid.khz_x);

			// MPI
			const int ipx = px[mode][side];
			const int dst = (ipx * Npy * Npz) + (Ipy * Npz) + Ipz;
			MPI_Sendrecv(Sendbuf_hy_x, Bid.numhy_x, MPI_REAL_T, dst, tag,
			             Recvbuf_hy_x, Bid.numhy_x, MPI_REAL_T, dst, tag, MPI_COMM_WORLD, &status);
			MPI_Sendrecv(Sendbuf_hz_x, Bid.numhz_x, MPI_REAL_T, dst, tag,
			             Recvbuf_hz_x, Bid.numhz_x, MPI_REAL_T, dst, tag, MPI_COMM_WORLD, &status);

			// buffer to H
			i = irecv[side];
			buffer_to_h(i, Recvbuf_hy_x, Hy, Bid.jhy_x, Bid.khy_x);
			buffer_to_h(i, Recvbuf_hz_x, Hz, Bid.jhz_x, Bid.khz_x);
		}
	}
#else
	mode = 0;  // dummy
#endif
}
